lib/TippyTopProvider.js

```diff
@@ -0,0 +1,67 @@
+/* globals require, JSON */
+"use strict";
+
+const data = require("sdk/self").data;
+const URL = require("sdk/url").URL;
+
+const DEFAULT_OPTIONS = {
+  sites: null
+};
+
+function TippyTopProvider(options = {}) {
+  this.options = Object.assign({}, DEFAULT_OPTIONS, options);
+  this.init();
+}
+
+TippyTopProvider.prototype = {
+  /**
+   * Initialize the sites object
+   */
+  init() {
+    // If the sites were passed into the constructor (for testability), use those.
+    // Else, read them from top_sites.json.
+    let sites;
+    if (this.options.sites) {
+      sites = this.options.sites;
+    } else {
+      sites = JSON.parse(data.load("content/favicons/top_sites.json"));
+    }
+
+    this._sitesByDomain = {};
+    sites.forEach(site => {
+      this._sitesByDomain[this._getKey(site.url)] = site;
+    });
+  },
+
+  /**
+    * Process the site, adding tippy top favicon and background color if URL is known.
+    */
+  processSite(site) {
+    let key;
+    try {
+      key = this._getKey(site.url);
+    } catch (e) {
+      key = null;
+    }
+
+    if (key && key in this._sitesByDomain) {
+      let tippyTopSite = this._sitesByDomain[key];
+      site.favicon_url = data.url("content/favicons/images/" + tippyTopSite.image_url);
+      site.background_color = tippyTopSite.background_color;
+    }
+    return site;
+  },
+
+  /**
+    * Create a key using the domain (minus the www.).
+    */
+  _getKey(url) {
+    let domain = URL(url).host;
```

Comment:
Yeah, I had talked about this with @oyiptong and we decided it was safer to match the full domain. I don't think there are that many cases in the top sites where we want to include all it's subdomains? I could be wrong.